File: Question 0695 - Max Area of Island/src/Solution.java
class Solution {
    private int dfs(int[][] grid, boolean[][] visited, int r, int c) {
        if (r >= 0 && r < grid.length && c >= 0 && c < grid[0].length && grid[r][c] == 1 && !visited[r][c]) {
            visited[r][c] = true;
            int up = dfs(grid, visited, r + 1, c);
            int down = dfs(grid, visited, r - 1, c);
            int right = dfs(grid, visited, r, c + 1);
            int left = dfs(grid, visited, r, c - 1);
            return 1 + up + down + right + left;
        }
        return 0;
    }

    public int maxAreaOfIsland(int[][] grid) {
        if (grid == null)
            return 0;
        int r = grid.length;
        int c = grid[0].length;

        boolean[][] visited = new boolean[r][c];

        int ans = 0;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (!visited[i][j] && grid[i][j] == 1)
                    ans = Math.max(ans, dfs(grid, visited, i, j));
            }
        }

        return ans;
    }
}

File: Question 0695 - Max Area of Island/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    int dfs(vector<vector<int>> &grid, vector<vector<int>> &dir, int r, int c)
    {
        if (r < 0 || r >= grid.size() || c < 0 || c >= grid[0].size() || grid[r][c] == 0)
            return 0;

        int ans = 1;
        grid[r][c] = 0;
        for (auto &d : dir)
        {
            int x = r + d[0];
            int y = c + d[1];

            ans += dfs(grid, dir, x, y);
        }

        return ans;
    }

public:
    int maxAreaOfIsland(vector<vector<int>> &grid)
    {
        int n = grid.size(), m = grid[0].size();

        int ans = INT_MIN;
        vector<vector<int>> dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (grid[i][j] == 1)
                    ans = max(ans, dfs(grid, dir, i, j));
            }
        }

        return ans == INT_MIN ? 0 : ans;
    }
};